It looks like the "build commits" job is failing because one of the commits is failing to build:

```
# github.com/cilium/cilium/daemon/cmd
Error: cmd/ciliumendpoints.go:103:22: undefined: resiliency.NewRetryableErr
Error: cmd/ciliumendpoints.go:130:21: undefined: resiliency.NewRetryableErr
```

Basically, each commit should be able to build independently.

--- a/daemon/cmd/ciliumendpoints.go
+++ b/daemon/cmd/ciliumendpoints.go
@@ -100,7 +100,7 @@ func (d *Daemon) deleteCiliumEndpoint(
 				return nil
 			}
 			logwf.WithError(err).Error("Failed to get possibly stale ciliumendpoints from apiserver")
-			return resiliency.NewRetryableErr(err)
+			return resiliency.Retryable(err)
 		}
 		if cep.Status.Networking.NodeIP != node.GetCiliumEndpointNodeIP() {
 			logwf.WithError(err).Debug("Stale CEP fetched apiserver no longer references this Node, skipping.")
@@ -127,7 +127,7 @@ func (d *Daemon) deleteCiliumEndpoint(
 			return nil
 		}
 		logwf.Error("Could not delete stale CEP")
-		return resiliency.NewRetryableErr(err)
+		return resiliency.Retryable(err)
 	}
 
 	return nil

--- a/daemon/cmd/daemon.go
+++ b/daemon/cmd/daemon.go
@@ -87,6 +87,7 @@ import (
 	"github.com/cilium/cilium/pkg/rate"
 	"github.com/cilium/cilium/pkg/recorder"
 	"github.com/cilium/cilium/pkg/redirectpolicy"
+	"github.com/cilium/cilium/pkg/resiliency"
 	"github.com/cilium/cilium/pkg/service"
 	serviceStore "github.com/cilium/cilium/pkg/service/store"
 	"github.com/cilium/cilium/pkg/source"
@@ -312,7 +313,7 @@ func createPrefixLengthCounter() *counter.PrefixLengthCounter {
 // that the most up-to-date information has been retrieved. At this point, the
 // daemon is aware of all the necessary information to restore the appropriate
 // IP.
-func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) error {
 	var (
 		cidrs  []*cidr.CIDR
 		fromFS net.IP
@@ -343,12 +344,7 @@ func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
 	}
 
 	restoredIP := node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidrs)
-	if err := removeOldRouterState(ipv6, restoredIP); err != nil {
-		log.WithError(err).Warnf(
-			"Failed to remove old router IPs (restored IP: %s) from cilium_host. Manual intervention is required to remove all other old IPs.",
-			restoredIP,
-		)
-	}
+	return removeOldRouterState(ipv6, restoredIP)
 }
 
 // removeOldRouterState will try to ensure that the only IP assigned to the
@@ -360,8 +356,9 @@ func removeOldRouterState(ipv6 bool, restoredIP net.IP) error {
 		// There's no old state remove as the host device doesn't exist and
 		// there's no restored IP anyway.
 		return nil
-	} else if err != nil {
-		return err
+	}
+	if err != nil {
+		return resiliency.Retryable(err)
 	}
 
 	family := netlink.FAMILY_V4
@@ -370,7 +367,7 @@ func removeOldRouterState(ipv6 bool, restoredIP net.IP) error {
 	}
 	addrs, err := netlink.AddrList(l, family)
 	if err != nil {
-		return err
+		return resiliency.Retryable(err)
 	}
 
 	isRestoredIP := func(a netlink.Addr) bool {
@@ -382,20 +379,16 @@ func removeOldRouterState(ipv6 bool, restoredIP net.IP) error {
 
 	log.Info("More than one stale router IP was found on the cilium_host device after restoration, cleaning up old router IPs.")
 
-	var errs []error
 	for _, a := range addrs {
 		if isRestoredIP(a) {
 			continue
 		}
-		if err := netlink.AddrDel(l, &a); err != nil {
-			errs = append(errs, fmt.Errorf("failed to remove IP %s: %w", a.IP, err))
+		if e := netlink.AddrDel(l, &a); e != nil {
+			err = errors.Join(err, resiliency.Retryable(fmt.Errorf("failed to remove IP %s: %w", a.IP, e)))
 		}
 	}
-	if len(errs) > 0 {
-		return fmt.Errorf("failed to remove all old router IPs: %v", errs)
-	}
 
-	return nil
+	return err
 }
 
 // newDaemon creates and returns a new Daemon with the parameters set in c.
@@ -1068,11 +1061,22 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 	// router IPs from the K8s resources if we weren't able to restore them
 	// from the fs. We must do this after IPAM because we must wait until the
 	// K8s resources have been synced. Part 2/2 of restoration.
-	if option.Config.EnableIPv4 {
-		d.restoreCiliumHostIPs(false, router4FromK8s)
+	gcHostIPsFn := func(ctx context.Context, retries int) (done bool, err error) {
+		var errs error
+		if option.Config.EnableIPv4 {
+			errs = errors.Join(errs, d.restoreCiliumHostIPs(false, router4FromK8s))
+		}
+		if option.Config.EnableIPv6 {
+			errs = errors.Join(errs, d.restoreCiliumHostIPs(true, router6FromK8s))
+		}
+		if resiliency.IsRetryable(errs) {
+			log.WithField(logfields.Attempt, retries).WithError(errs).Warnf("Failed to remove old router IPs from cilium_host.")
+			return false, nil
+		}
+		return true, errs
 	}
-	if option.Config.EnableIPv6 {
-		d.restoreCiliumHostIPs(true, router6FromK8s)
+	if err := resiliency.Retry(ctx, 100*time.Millisecond, 3, gcHostIPsFn); err != nil {
+		log.WithError(err).Error("Restore of the cilium_host ips failed. Manual intervention is required to remove all other old IPs.")
 	}
 
 	// restore endpoints before any IPs are allocated to avoid eventual IP

--- a/pkg/resiliency/error.go
+++ b/pkg/resiliency/error.go
@@ -3,12 +3,12 @@
 
 package resiliency
 
-// RetryableErr tracks errors that could be retried.
-type RetryableErr struct {
+// retryableErr tracks errors that could be retried.
+type retryableErr struct {
 	error
 }
 
-// NewRetryableErr returns a new instance.
-func NewRetryableErr(e error) RetryableErr {
-	return RetryableErr{error: e}
+// Retryable returns a new instance.
+func Retryable(e error) retryableErr {
+	return retryableErr{error: e}
 }

--- a/pkg/resiliency/helpers.go
+++ b/pkg/resiliency/helpers.go
@@ -1,0 +1,32 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package resiliency
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// RetryFunc tracks resiliency retry calls.
+type RetryFunc func(ctx context.Context, retries int) (bool, error)
+
+// Retry retries the provided call using exponential retries given an initial duration for up to max retries count.
+func Retry(ctx context.Context, duration time.Duration, maxRetries int, fn RetryFunc) error {
+	bo := wait.Backoff{
+		Duration: duration,
+		Factor:   1,
+		Jitter:   0.1,
+		Steps:    maxRetries,
+	}
+
+	var retries int
+	f := func(ctx context.Context) (bool, error) {
+		retries++
+		return fn(ctx, retries)
+	}
+
+	return wait.ExponentialBackoffWithContext(ctx, bo, f)
+}

--- a/pkg/resiliency/helpers_test.go
+++ b/pkg/resiliency/helpers_test.go
@@ -1,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package resiliency_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/cilium/cilium/pkg/resiliency"
+)
+
+func TestRetries(t *testing.T) {
+	var maxRetries int
+
+	uu := map[string]struct {
+		d   time.Duration
+		m   int
+		f   resiliency.RetryFunc
+		err error
+		e   int
+	}{
+		"no-retries": {
+			d: 10 * time.Millisecond,
+			m: 3,
+			f: func(ctx context.Context, retries int) (bool, error) {
+				maxRetries = retries
+				return true, nil
+			},
+			e: 1,
+		},
+		"happy": {
+			d: 10 * time.Millisecond,
+			m: 3,
+			f: func(ctx context.Context, retries int) (bool, error) {
+				maxRetries = retries
+				if retries < 3 {
+					return false, nil
+				}
+				return true, nil
+			},
+			e: 3,
+		},
+		"error-complete": {
+			d: 10 * time.Millisecond,
+			m: 3,
+			f: func(ctx context.Context, retries int) (bool, error) {
+				maxRetries = retries
+				return true, errors.New("boom")
+			},
+			err: errors.New("boom"),
+			e:   1,
+		},
+		"error-retry": {
+			d: 10 * time.Millisecond,
+			m: 3,
+			f: func(ctx context.Context, retries int) (bool, error) {
+				maxRetries = retries
+				return false, errors.New("boom")
+			},
+			err: errors.New("boom"),
+			e:   1,
+		},
+	}
+
+	for k := range uu {
+		u := uu[k]
+		maxRetries = 0
+		t.Run(k, func(t *testing.T) {
+			err := resiliency.Retry(context.Background(), u.d, u.m, u.f)
+			if err != nil {
+				assert.Equal(t, u.err, err)
+			}
+			assert.Equal(t, u.e, maxRetries)
+		})
+	}
+}

--- a/pkg/resiliency/retry.go
+++ b/pkg/resiliency/retry.go
@@ -9,5 +9,5 @@ import (
 
 // IsRetryable checks if an error can be retried.
 func IsRetryable(e error) bool {
-	return errors.As(e, new(RetryableErr))
+	return errors.As(e, new(retryableErr))
 }

